sqlstats: set default value of sql.stats.flush.minimum_interval to 10m

xinhaoz opened the pull request:
**sqlstats: cleanup flush, move min flush interval check**

This commit moves the check previously in `MaybeFlush` that
aborts the flush if not enough time has passed since the
last flush operation began to the sql stats flush loop.

`MaybeFlush` should be called by other functions to trigger
a flush of sql stats immediately (e.g. on server drain). The
minumum flush interval is only relevant to the sql stats flush
worker. Other checks, such as aborting due to flush being
disabled or due to the maximum number of rows being reached,
remain with `MaybeFlush`.

Epic: none
Part of: #123091

Release note: None

**sqlstats: add metric tracking skipped sql stats flushes**

This commit adds the metric `sql.stats.flushes.skipped.count`
which tracks the number of flushes skipped (triggered but
not started).

Epic: none
Part of : #123091

Release note (ops change): New metric, `sql.stats.flushes.skipped.count`
tracks the number of sql stats flushes that were skipped either due to
not meeting the minimum flush duration or the system tables being full.

**sqlstats: set default value of sql.stats.flush.minimum_interval to 10m**

This commit sets the minimum flush interval for sql stats to be
equivalent to the default flush interval, 10m. This ensures that
the flush will not be allowed to trigger consecutively when there
is constant memory pressure (ex. due to hitting in-memory fingerprint
limit) on the sql stats system. We want to prevent the flush from
constantly running in the BG since it can use up resources.

Fixes: https://github.com/cockroachdb/cockroach/issues/123091
Epic: CRDB-37544

xinhaoz committed:
sqlstats: set default value of sql.stats.flush.minimum_interval to 10m

This commit sets the minimum flush interval for sql stats to be
equivalent to the default flush interval, 10m. This ensures that
the flush will not be allowed to trigger consecutively when there
is constant memory pressure (ex. due to hitting in-memory fingerprint
limit) on the sql stats system. We want to prevent the flush from
constantly running in the BG since it can use up resources.

Fixes: #123091
Epic: CRDB-37544

Release note: None

xinhaoz committed:
sqlstats: cleanup flush, move min flush interval check

This commit moves the check previously in `MaybeFlush` that
aborts the flush if not enough time has passed since the
last flush operation began to the sql stats flush loop.

`MaybeFlush` should be called by other functions to trigger
a flush of sql stats immediately (e.g. on server drain). The
minumum flush interval is only relevant to the sql stats flush
worker. Other checks, such as aborting due to flush being
disabled or due to the maximum number of rows being reached,
remain with `MaybeFlush`.

Epic: none
Part of: #123091

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/123466)
<!-- Reviewable:end -->

xinhaoz committed:
sqlstats: add metric tracking skipped sql stats flushes

This commit adds the metric `sql.stats.flushes.skipped.count`
which tracks the number of flushes skipped (triggered but
not started).

Epic: none
Part of : #123091

Release note (ops change): New metric, `sql.stats.flushes.skipped.count`
tracks the number of sql stats flushes that were skipped either due to
not meeting the minimum flush duration or the system tables being full.

xinhaoz commented:
As per discussion, opened https://github.com/cockroachdb/cockroach/pull/123567 with just the CS change. I'll modify this PR as a follow-up to that.

xinhaoz commented:
@dhartunian Reviving this PR and closed the one linked above. I omited the small portion that was technically a bug fix, but I think all these other changes can be batched together. Thoughts?

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just some nits

> Reviewed 1 of 2 files at r3, 7 of 7 files at r4, 2 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/123466#-:-NxDVoK0CMCHSCEiLAk5:b-yl2rbz)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @nkodali, and @xinhaoz)

___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 376 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/123466#-NxDUYPgAK8J7Zkri1IV:-NxDUYPgAK8J7Zkri1IW:bnzwxt3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf5052b134dafb6d6c5afb2214d923665decd06d/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L376)):*
> ```Go
> 
> 	testutils.SucceedsSoon(t, func() error {
> 		if srv.SQLServer().(*sql.Server).ServerMetrics.StatsMetrics.SQLStatsFlushesSkipped.Count() < 0 {
> ```

can you add confirmation prior to the execution of above statements, that skipped count is zero?
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 402 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/123466#-NxDUiJB3Uzs-wjTeIsF:-NxDUiJB3Uzs-wjTeIsG:b-6d4g54) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf5052b134dafb6d6c5afb2214d923665decd06d/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L402)):*
> ```Go
> 	srv, conn, _ := serverutils.StartServer(t, base.TestServerArgs{
> 		Knobs: base.TestingKnobs{
> 			SQLStatsKnobs: sqlstats.CreateTestingKnobs(),
> ```

does this do anything?


<!-- Sent from Reviewable.io -->
